### PR TITLE
applications: nrf_desktop: Switch to new workqueue API

### PR DESCRIPTION
**Pull request by pdunaj**
Jira:NCSDK-9400

Signed-off-by: Pawel Dunaj <pawel.dunaj@nordicsemi.no>

### CODE REVIEW

**Review by MarekPieta — APPROVED** [review]

**Comment by pdunaj** [comment]
rebase + retrigger CI

**Comment by pdunaj** [comment]
rebased for CI rerun

File: applications/nrf_desktop/src/hw_interface/battery_charger.c
Comment (pabigot) [inline applications/nrf_desktop/src/hw_interface/battery_charger.c:246]: If nrf-sdk hasn't adopted the policy of documenting why it's safe to do unchecked work cancels it should.

File: applications/nrf_desktop/src/modules/ble_scan.c
Comment (pabigot) [inline applications/nrf_desktop/src/modules/ble_scan.c:396]: FWIW I believe this can return negative only if `SCAN_TRIG_CHECK_MS` is zero; is that a legitimate configuration?

See also https://github.com/zephyrproject-rtos/zephyr/pull/35337 inspired by this idiom.

**Review by pabigot — CHANGES_REQUESTED** [review]
Unless I'm wrong that (re)schedule can fail only if `K_NO_WAIT` is passed the assert checks on its return value should be removed when the delay is positive.

It would be good to document why it's safe to use `k_work_cancel_delayable` without checking the error return, but that's not blocking if nrf-sdk hasn't adopted that recommended policy.

File: applications/nrf_desktop/src/hw_interface/battery_charger.c
Comment (pdunaj) [reply applications/nrf_desktop/src/hw_interface/battery_charger.c:246]: This cannot fail by design as cancel is done from the system workqueue context. The thing is I would have to add it at almost every cancel instance that will not help but add noise instead.

File: applications/nrf_desktop/src/modules/ble_scan.c
Comment (pdunaj) [reply applications/nrf_desktop/src/modules/ble_scan.c:396]: you are actually right; check here has little sense as `SCAN_TRIG_CHECK_MS` cannot be zero (we validate that)

Comment (pdunaj) [reply applications/nrf_desktop/src/modules/ble_scan.c]: removed the unnecessary checks

File: applications/nrf_desktop/src/hw_interface/battery_charger.c
Comment (pdunaj) [reply applications/nrf_desktop/src/hw_interface/battery_charger.c:246]: If needed I would prefer to add a note in README.

Comment (pdunaj) [reply applications/nrf_desktop/src/hw_interface/battery_charger.c:246]: But to be honest I think it has more sense to add any comment only if we suspect something can fail but we don't bother or simply handle it.
The `k_work_cancel_delayable` should not be threated any way special.

Will add assert.

Comment (pdunaj) [reply applications/nrf_desktop/src/hw_interface/battery_charger.c:246]: Would add assert but because of checkpatch i would  have to add 3 lines of code instead of 1.
instead of 
```
int err = stuff();
assert(!err);
```

it will make me add 
```
int err = stuff();

assert(!err);
unused(err);
```

this will be even worse then adding a comment.

Comment (pdunaj) [reply applications/nrf_desktop/src/hw_interface/battery_charger.c:246]: Hi @carlescufi , please tell if I should add a comment or entry in README is enough, or I can simply ignore this.

Comment (carlescufi) [reply applications/nrf_desktop/src/hw_interface/battery_charger.c:246]: I fear that our customers will copy code from these reference applications without being aware of the pitfalls. I would rather use ASSERT than anything else, because that will make technical support requests much simpler to deal with: The assertion will fail in the user's code, giving us a lead.

Comment (pabigot) [reply applications/nrf_desktop/src/hw_interface/battery_charger.c:246]: To repeat my opinion:

> If nrf-sdk hasn't adopted the policy of documenting why it's safe to do unchecked work cancels it should.

It's entirely up to y'all, but [upstream specifically recommends](https://docs.zephyrproject.org/latest/reference/kernel/threads/workqueue.html#check-return-values) a comment so people who look at (and copy) the code are (a) reminded that something might go wrong, and (b) not obligated to completely analyze the code to determine whether a failure to successfully cancel explains why their application is unreliable.

Remember that asserts do get turned off sometimes, at which point they are useless (especially if there's no comment explaining *why* something "can't happen").

**Review by pabigot — APPROVED** [review]

**Comment by pabigot** [comment]
Approved since the unnecessary return code checks have been removed.  Whether the absence of explanatory comments should block merge is outside my purview.

**Review by trond-snekvik — APPROVED** [review]

File: applications/nrf_desktop/src/hw_interface/battery_charger.c
Comment (pdunaj) [reply applications/nrf_desktop/src/hw_interface/battery_charger.c:246]: @carlescufi can we merge it as is? it seems that other PRs regarding api change are merged without such comments or asserts.

Comment (carlescufi) [reply applications/nrf_desktop/src/hw_interface/battery_charger.c:246]: I would really like to document this or assert, which can serve as documentation. I understand we haven't been strict in other cases, but this is a reference application, so I would like to be a bit more strict here. I think we need to go back and document the uses of `k_work_cancel_delayable()` that are not checked.

**Review by trond-snekvik — CHANGES_REQUESTED** [review]
NACKing until Carles' request has been addressed

File: applications/nrf_desktop/src/hw_interface/battery_charger.c
Comment (carlescufi) [reply applications/nrf_desktop/src/hw_interface/battery_charger.c:246]: @pabigot can you point to us a reference commented instance upstream that we can use as inspiration?

Comment (pdunaj) [reply applications/nrf_desktop/src/hw_interface/battery_charger.c:246]: I can add the following
```
/* k_work_cancel_delayable cannot fail when called from the context of another work. */
```
I don't see the purpose of adding information about general problems with canceling work.
There are plenty of API calls where return code is ignored as it is expected not to fail. Why we threat `k_work_cancel_delayable` differently I don't understand.

I would have no problem with adding an assert (as actually I think this is the best thing to add here). But there is a ton of checkpatch overhead related to such a minor change that I just simply hate (having mess added in the code only to appease a tool).

(Why it cannot be a simple one liner - why we need to have empty line after every single variable declaration. Why when asserts are removed there is nothing to appease the compiler...)

Comment (pabigot) [reply applications/nrf_desktop/src/hw_interface/battery_charger.c:246]: @carlescufi 

> @pabigot can you point to us a reference commented instance upstream that we can use as inspiration?

https://docs.zephyrproject.org/latest/reference/kernel/threads/workqueue.html#check-return-values shows one that comes from one of @trond-snekvik's commits.

Comment (pabigot) [reply applications/nrf_desktop/src/hw_interface/battery_charger.c:246]: > There are plenty of API calls where return code is ignored as it is expected not to fail. Why we threat `k_work_cancel_delayable` differently I don't understand.

Because historically people used it inappropriately, assumed it wouldn't fail (e.g. assumed cooperative scheduling on a uniprocessor), and failed to account for cases where it did, resulting in subtle race conditions when things changed or code was copied into a context where the assumption of not failing was no longer valid.  There are years of mis-use and habit to overcome.  A comment assures future maintainers of the assumptions made, and is IMO a low burden to impose relative to the cost of repeated reanalysis and diagnosis.

I did personally sometimes accept `(void)func();` as an explicit indication that the developer was intentionally ignoring a return value because it wasn't relevant, though upstream move to MISRA "compliance" means that idiom might have been added without any analysis just to pass CI, so it's meaningless.

Comment (pdunaj) [reply applications/nrf_desktop/src/hw_interface/battery_charger.c]: added a comment

**Review by trond-snekvik — APPROVED** [review]

**Pull request merged**
